File: CodeExchange/cx.go
package main

// CodeExchange (cx) v3.0.0 — Go Implementation
// High-performance streaming layout utilizing native io.Writer constructs.

import (
	"fmt"
	"io"
	"os"
	"strconv"
)

const ioBufferSize = 16384
const hexDigits = "0123456789ABCDEF"

func emitRawBuffer(output io.Writer, buffer []byte) {
	if len(buffer) == 0 {
		return
	}
	var crc byte = 0
	for _, b := range buffer {
		crc ^= b
	}
	output.Write(buffer)
	fmt.Fprintf(output, "|%c%c\n", hexDigits[(crc>>4)&0x0F], hexDigits[crc&0x0F])
}

func flushRun(buffer []byte, charByte byte, count int, output io.Writer, lineLength int) []byte {
	if count == 0 {
		return buffer
	}

	var normalSeq []byte
	switch charByte {
	case '\n': normalSeq = []byte("\\n")
	case '\t': normalSeq = []byte("\\t")
	case '\r': normalSeq = []byte("\\r")
	case '\\': normalSeq = []byte("\\\\")
	default:   normalSeq = []byte{charByte}
	}

	normalWidth := len(normalSeq)
	maxAllowed := lineLength - 4
	if lineLength <= 4 {
		maxAllowed = lineLength
	}

	for count > 0 {
		chunk := count
		if chunk > 65535 {
			chunk = 65535
		}
		rawCost := normalWidth * chunk
		rleCost := 4 + normalWidth
		if chunk > 255 {
			rleCost = 6 + normalWidth
		}

		if lineLength > 0 && (len(buffer)+rleCost > maxAllowed) {
			if len(buffer) > 0 {
				emitRawBuffer(output, buffer)
				buffer = buffer[:0]
				continue
			}
		}

		if rleCost < rawCost {
			var marker string
			if chunk > 255 {
				marker = fmt.Sprintf("/z%04X", chunk)
			} else {
				marker = fmt.Sprintf("/x%02X", chunk)
			}
			buffer = append(buffer, []byte(marker)...)
			buffer = append(buffer, normalSeq...)
			count -= chunk
		} else {
			processNow := chunk
			if lineLength > 0 {
				spaceLeft := maxAllowed - len(buffer)
				if spaceLeft < 0 {
					spaceLeft = 0
				}
				maxRaw := spaceLeft / normalWidth
				if maxRaw == 0 {
					emitRawBuffer(output, buffer)
					buffer = buffer[:0]
					continue
				}
				if processNow > maxRaw {
					processNow = maxRaw
				}
			}
			for i := 0; i < processNow; i++ {
				buffer = append(buffer, normalSeq...)
			}
			count -= processNow
		}

		if lineLength > 0 && len(buffer) >= maxAllowed {
			emitRawBuffer(output, buffer)
			buffer = buffer[:0]
		}
	}
	return buffer
}

func encodeFile(output io.Writer, filePath string, lineLength int) {
	file, err := os.Open(filePath)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error opening file: %v\n", err)
		return
	}
	defer file.Close()

	fmt.Fprintf(output, "===== FILE: %s - LL:%d ====\n", filePath, lineLength)

	var globalCrc byte = 0
	var currentByte byte = 0
	runCount := 0
	lineBuffer := make([]byte, 0, 1024)
	ioBuf := make([]byte, ioBufferSize)

	for {
		bytesRead, err := file.Read(ioBuf)
		if bytesRead > 0 {
			for i := 0; i < bytesRead; i++ {
				b := ioBuf[i]
				globalCrc ^= b

				if runCount == 0 {
					currentByte = b
					runCount = 1
				} else if b == currentByte {
					runCount++
				} else {
					lineBuffer = flushRun(lineBuffer, currentByte, runCount, output, lineLength)
					currentByte = b
					runCount = 1
				}
			}
		}
		if err == io.EOF {
			break
		}
	}

	if runCount > 0 {
		lineBuffer = flushRun(lineBuffer, currentByte, runCount, output, lineLength)
	}
	if len(lineBuffer) > 0 {
		emitRawBuffer(output, lineBuffer)
	}

	fmt.Fprintf(output, "===== ENDFILE: %s - FCS:%02X ====\n", filePath, globalCrc)
}

func main() {
	args := os.Args
	if len(args) < 2 {
		return
	}

	if args[1] == "-e" {
		lineLimit := 80
		fileIdx := 2
		if len(args) > 3 && args[2] == "-l" {
			if val, err := strconv.Atoi(args[3]); err == nil {
				lineLimit = val
			}
			fileIdx = 4
		}
		fmt.Printf("CodeExchange format - xc format v3 (Single-Pass Stream Architecture)\n\n")
		for i := fileIdx; i < len(args); i++ {
			encodeFile(os.Stdout, args[i], lineLimit)
		}
		fmt.Printf("===== EOF\n")
	}
}

File: CodeExchange/cx.js
#!/usr/bin/env node
/**
 * CodeExchange (cx) v3.0.0 — Node.js Implementation
 * High-performance streaming architecture using file descriptor blocks.
 */

const fs = require('fs');
const path = require('path');

const HEX_DIGITS = "0123456789ABCDEF";
const IO_BUFFER_SIZE = 16384;

function emitRawBuffer(outputStream, buffer) {
    if (buffer.length === 0) return;
    let crc = 0;
    for (let i = 0; i < buffer.length; i++) {
        crc ^= buffer[i];
    }
    const hexCrc = HEX_DIGITS[(crc >> 4) & 0x0F] + HEX_DIGITS[crc & 0x0F];
    outputStream.write(buffer.toString('ascii') + `|${hexCrc}\n`);
}

function flushRun(buffer, charByte, count, outputStream, lineLength) {
    if (count === 0) return buffer;

    let normalSeq;
    switch (charByte) {
        case 10: normalSeq = Buffer.from('\\n'); break;
        case 9:  normalSeq = Buffer.from('\\t'); break;
        case 13: normalSeq = Buffer.from('\\r'); break;
        case 92: normalSeq = Buffer.from('\\\\'); break;
        default: normalSeq = Buffer.from([charByte]);
    }
    
    const normalWidth = normalSeq.length;
    const maxAllowed = lineLength > 4 ? lineLength - 4 : lineLength;

    while (count > 0) {
        const chunk = Math.min(count, 65535);
        const rawCost = normalWidth * chunk;
        const rleCost = chunk > 255 ? (6 + normalWidth) : (4 + normalWidth);

        if (lineLength > 0 && (buffer.length + rleCost > maxAllowed)) {
            if (buffer.length > 0) {
                emitRawBuffer(outputStream, buffer);
                buffer = Buffer.alloc(0);
                continue;
            }
        }

        if (rleCost < rawCost) {
            const markerStr = chunk > 255 
                ? `/z${chunk.toString(16).toUpperCase().padStart(4, '0')}`
                : `/x${chunk.toString(16).toUpperCase().padStart(2, '0')}`;
            buffer = Buffer.concat([buffer, Buffer.from(markerStr), normalSeq]);
            count -= chunk;
        } else {
            let processNow = chunk;
            if (lineLength > 0) {
                const spaceLeft = Math.max(0, maxAllowed - buffer.length);
                const maxRaw = Math.floor(spaceLeft / normalWidth);
                if (maxRaw === 0) {
                    emitRawBuffer(outputStream, buffer);
                    buffer = Buffer.alloc(0);
                    continue;
                }
                processNow = Math.min(processNow, maxRaw);
            }

            const repeated = Buffer.alloc(processNow * normalWidth);
            for (let i = 0; i < processNow; i++) {
                normalSeq.copy(repeated, i * normalWidth);
            }
            buffer = Buffer.concat([buffer, repeated]);
            count -= processNow;
        }

        if (lineLength > 0 && buffer.length >= maxAllowed) {
            emitRawBuffer(outputStream, buffer);
            buffer = Buffer.alloc(0);
        }
    }
    return buffer;
}

function encodeFile(outputStream, filePath, lineLength) {
    if (!fs.existsSync(filePath)) {
        process.stderr.write(`Error: File not found: ${filePath}\n`);
        return;
    }

    outputStream.write(`===== FILE: ${filePath} - LL:${lineLength} ====\n`);
    
    let globalCrc = 0;
    let lineBuffer = Buffer.alloc(0);
    let currentByte = null;
    let runCount = 0;

    const fd = fs.openSync(filePath, 'r');
    const ioBuf = Buffer.alloc(IO_BUFFER_SIZE);
    let bytesRead = 0;

    while ((bytesRead = fs.readSync(fd, ioBuf, 0, IO_BUFFER_SIZE, null)) > 0) {
        for (let i = 0; i < bytesRead; i++) {
            const b = ioBuf[i];
            globalCrc ^= b;

            if (runCount === 0) {
                currentByte = b;
                runCount = 1;
            } else if (b === currentByte) {
                runCount++;
            } else {
                lineBuffer = flushRun(lineBuffer, currentByte, runCount, outputStream, lineLength);
                currentByte = b;
                runCount = 1;
            }
        }
    }
    fs.closeSync(fd);

    if (runCount > 0) {
        lineBuffer = flushRun(lineBuffer, currentByte, runCount, outputStream, lineLength);
    }
    if (lineBuffer.length > 0) {
        emitRawBuffer(outputStream, lineBuffer);
    }

    const hexGlobal = globalCrc.toString(16).toUpperCase().padStart(2, '0');
    outputStream.write(`===== ENDFILE: ${filePath} - FCS:${hexGlobal} ====\n`);
}

// CLI Command Route Dispatcher
const args = process.argv.slice(2);
if (args[0] === '-e') {
    let lineLimit = 80;
    let fileIdx = 1;
    if (args[1] === '-l') {
        lineLimit = parseInt(args[2], 10) || 80;
        fileIdx = 3;
    }
    process.stdout.write("CodeExchange format - xc format v3 (Single-Pass Stream Architecture)\n\n");
    for (let i = fileIdx; i < args.length; i++) {
        encodeFile(process.stdout, args[i], lineLimit);
    }
    process.stdout.write("===== EOF\n");
}
